OpenSim/Common/Reporter.h

```diff
@@ -0,0 +1,252 @@
+*/
+template<typename InputsT = SimTK::Real>
+class Reporter : public AbstractReporter {
+    OpenSim_DECLARE_ABSTRACT_OBJECT_T(Reporter, InputsT, AbstractReporter);
+public:
+    //=========================================================================
+    // INPUTS
+    //=========================================================================
+    /** Variable list of Inputs by name that will be recorded by the Reporter
+    and corresponds to a Model (Component) Output during a simulation. */
+    OpenSim_DECLARE_LIST_INPUT(inputs, InputsT, SimTK::Stage::Report,
+        "Variable list of inputs to be reported.");
+
+protected:
+    /** Default constructor sets up Reporter-level properties; can only be
+    called from a derived class constructor. **/
+    Reporter() = default;
+    virtual ~Reporter() = default;
+    //=============================================================================
+};  // END of class Reporter<InputsT>
+    //=============================================================================
+
+
+/**
+* This concrete Reporter class collects and reports Output<T>s within a 
+* TimeSeriesTable. The column labels are the Output names and the contents are
+* the Output values with each row being the value of all outputs at subsequent
+* times determined by the reporting interval.
+*
+* @author Ajay Seth
+*/
+template<class InputT=Real, typename ValueT=Real>
+class TableReporter : public Reporter<InputT> {
+OpenSim_DECLARE_CONCRETE_OBJECT_T(TableReporter, InputT, Reporter<InputT>);
+public:
+    TableReporter() = default;
+    virtual ~TableReporter() = default;
+
+    const TimeSeriesTable_<ValueT>& getReport() const {
+        return _outputTable;
+    }
+
+protected:
+    void implementReport(const SimTK::State& state) const override {
+        const auto& input = getInput<InputT>("inputs");
+        SimTK::Vector_<ValueT> result(int(input.getNumConnectees()));
+
+        for (auto idx = 0; idx < input.getNumConnectees(); ++idx) {
+              const auto& chan = Input<InputT>::downcast(input).getChannel(idx);
+              const auto& value = chan.getValue(state);
+              result[idx] = value;
+        }
+        _outputTable.appendRow(state.getTime(), ~result);
+    }
+
+
+    void extendConnect(Component& root) override {
+        Super::extendConnect(root);
+
+        const auto& input = getInput<InputT>("inputs");
+
+        std::vector<std::string> labels;
+        for (auto idx = 0; idx < input.getNumConnectees(); ++idx) {
+            const auto& chan = Input<InputT>::downcast(input).getChannel(idx);
+            std::string label = chan.getName();
+            if (label.empty()) {
+                label = chan.getOutput().getName();
+            }
+            labels.push_back(label);
+        }
+        _outputTable.setColumnLabels(labels);
+    }
+
+private:
+
+    // Hold the output values in a table with values as columns and time rows
+    mutable TimeSeriesTable_<ValueT> _outputTable;
```

Comment:
I'll also add a comment describing why it's okay for this to be written to in a const method.